include/drives.h

```diff
@@ -75,13 +76,15 @@ class localDrive : public DOS_Drive {
 	virtual bool isRemovable(void);
 	virtual Bits UnMount(void);
 	const char* getBasedir() {return basedir;};
+	virtual bool isNewWriteProtectedFile(const std::string& filename);
 protected:
 	char basedir[CROSS_LEN];
 	struct {
 		char srch_dir[CROSS_LEN];
 	} srchInfo[MAX_OPENDIRS];
 
 private:
+	std::set<std::string> write_protected_files;
```

Comment:
I don't have any specific comment here, just flagging this as a potential source of problems. I had lots of problems with `std::set` usage in the past (especially with sets storing objects) - problems usually resulted from iterator invalidation, which can happen really easily can usually flies under the radar.

So it's a note for myself: double check usage of this set in this patch.

Reply:
My original code used a simple vector<string>, that the check function would iterator over.
It's performance was O(N) though and the hit would become additive as games with hundreds (maybe thousands of files) repeatedly re-open the same file.

set<string> drops this hit to log(N) and is technically the right data-structure for the job, however for small N, crawling a couple hundreds vector items is just as fast as set<>. 

So we could easily flip this back to vector<list> if you've hit real-world issues with <set>.

Reply:
Iterator invalidation can be guarded against with a strategy like Java's which increases a hidden var during adding and checks it during getting, which is safe single threaded and 'statistically safe' on multthhreads. Why doesn't c++ do this?

Reply:
The code only uses the set (and its content) atomically: we check if it contains a filename. If yes, we return a false bool, otherwise we add the filename to the set and return a true bool. We never carry out an iterator from the set. 

The purpose of the set is to simply reduce verbosity to the user - we print a message one time only for a given file that's protected.  When the DOS game opens the same file a second time, we skip the message. (Without this 'tracking', we would be flooding the console with duplicate message because some games rapidly cycles through the same files over and over)